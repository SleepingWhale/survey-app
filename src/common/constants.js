export const types = {
  string: 'string',
  number: 'number',
  boolean: 'boolean',
  date: 'date'
};
