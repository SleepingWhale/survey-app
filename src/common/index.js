export { Validators } from './utils';
export { types, validationMessages } from './constants';
