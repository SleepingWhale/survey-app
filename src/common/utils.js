import { types } from './constants';

export class Validators {
  static [types.string] = value =>
    (typeof value === 'string' || value instanceof String) && value.length > 0;

  static [types.date] = value => value instanceof Date;

  static [types.boolean] = value => typeof value === 'boolean';

  static [types.number] = value => {
    let _value = value;
    if (this.string(value) && /^-?\d*[.,]?\d*$/.test(value))
      _value = parseFloat(_value);
    return typeof _value === 'number' && Number.isFinite(_value) && _value > 0;
  };
}
