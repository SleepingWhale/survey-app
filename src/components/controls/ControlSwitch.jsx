import React from 'react';
import PropTypes from 'prop-types';
import { InputText } from './InputText';
import { InputNumber } from './InputNumber';
import { InputCheckbox } from './InputCheckbox';
import { InputDate } from './InputDate';
import { types } from '../../common';

const Controls = {
  [types.string]: InputText,
  [types.number]: InputNumber,
  [types.boolean]: InputCheckbox,
  [types.date]: InputDate
};

export const ControlSwitch = ({ type, ...rest }) => {
  const Control = Controls[type];

  if (!Control)
    return (
      <div className="form-group">
        <span className="badge badge-info">
          Something wrong with this question. Please, proceed to the next one.
        </span>
      </div>
    );

  return <Control {...rest} />;
};

ControlSwitch.prototype = {
  type: PropTypes.oneOf(Object.values(types))
};
