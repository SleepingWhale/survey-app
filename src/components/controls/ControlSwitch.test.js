import React from 'react';
import { mount } from 'enzyme';
import { ControlSwitch } from './ControlSwitch';
import { types } from '../../common';

describe('ControlSwitch', () => {
  it(`should render control for ${types.boolean} type`, () => {
    const onChangeHandler = jest.fn();
    const component = mount(
      <ControlSwitch
        onChange={onChangeHandler}
        value={false}
        id="id"
        question="question"
        type={types.boolean}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it(`should render control for ${types.string} type`, () => {
    const onChangeHandler = jest.fn();
    const component = mount(
      <ControlSwitch
        onChange={onChangeHandler}
        value="hello world"
        id="id"
        question="question"
        type={types.string}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it(`should render control for ${types.number} type`, () => {
    const onChangeHandler = jest.fn();
    const component = mount(
      <ControlSwitch
        onChange={onChangeHandler}
        value={55}
        id="id"
        question="question"
        type={types.number}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it(`should render control for ${types.date} type`, () => {
    const onChangeHandler = jest.fn();
    const component = mount(
      <ControlSwitch
        onChange={onChangeHandler}
        value={new Date(1563382002751)}
        id="id"
        question="question"
        type={types.date}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
