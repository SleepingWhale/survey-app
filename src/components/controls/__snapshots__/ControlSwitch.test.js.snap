// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlSwitch should render control for boolean type 1`] = `
<ControlSwitch
  id="id"
  onChange={[MockFunction]}
  question="question"
  type="boolean"
  value={false}
>
  <InputCheckbox
    id="id"
    onChange={[MockFunction]}
    question="question"
    value={false}
  >
    <div
      className="form-group"
    >
      <div
        className="form-check"
      >
        <input
          checked={false}
          className="form-check-input"
          id="id"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          className="form-check-label"
          htmlFor="id"
        >
          question
        </label>
      </div>
    </div>
  </InputCheckbox>
</ControlSwitch>
`;

exports[`ControlSwitch should render control for date type 1`] = `
<ControlSwitch
  id="id"
  onChange={[MockFunction]}
  question="question"
  type="date"
  value={2019-07-17T16:46:42.751Z}
>
  <InputDate
    id="id"
    onChange={[MockFunction]}
    question="question"
    value={2019-07-17T16:46:42.751Z}
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="id"
      >
        question
      </label>
      <br />
      <DatePicker
        allowSameDay={false}
        autoComplete="off"
        className="form-control"
        dateFormat="MM/dd/yyyy"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        id="id"
        inlineFocusSelectedMonth={false}
        locale="enGB"
        monthsShown={1}
        nextMonthButtonLabel="Next month"
        onBlur={[Function]}
        onChange={[MockFunction]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        readOnly={false}
        renderDayContents={[Function]}
        selected={2019-07-17T16:46:42.751Z}
        shouldCloseOnSelect={true}
        showMonthYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
      >
        <PopperComponent
          hidePopper={true}
          popperComponent={null}
          popperModifiers={
            Object {
              "preventOverflow": Object {
                "boundariesElement": "viewport",
                "enabled": true,
                "escapeWithReference": true,
              },
            }
          }
          popperPlacement="bottom-start"
          popperProps={Object {}}
          targetComponent={
            <div
              className="react-datepicker__input-container"
            >
              <input
                autoComplete="off"
                className="form-control"
                disabled={false}
                id="id"
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={false}
                type="text"
                value="07/17/2019"
              />
            </div>
          }
        >
          <Manager>
            <Reference>
              <InnerReference
                setReferenceNode={[Function]}
              >
                <div
                  className="react-datepicker-wrapper"
                >
                  <div
                    className="react-datepicker__input-container"
                  >
                    <input
                      autoComplete="off"
                      className="form-control"
                      disabled={false}
                      id="id"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      readOnly={false}
                      type="text"
                      value="07/17/2019"
                    />
                  </div>
                </div>
              </InnerReference>
            </Reference>
          </Manager>
        </PopperComponent>
      </DatePicker>
    </div>
  </InputDate>
</ControlSwitch>
`;

exports[`ControlSwitch should render control for number type 1`] = `
<ControlSwitch
  id="id"
  onChange={[MockFunction]}
  question="question"
  type="number"
  value={55}
>
  <Input
    id="id"
    onChange={[MockFunction]}
    question="question"
    value={55}
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="id"
      >
        question
      </label>
      <input
        autoComplete="off"
        className="form-control"
        id="id"
        onChange={[Function]}
        type="number"
        value={55}
      />
    </div>
  </Input>
</ControlSwitch>
`;

exports[`ControlSwitch should render control for string type 1`] = `
<ControlSwitch
  id="id"
  onChange={[MockFunction]}
  question="question"
  type="string"
  value="hello world"
>
  <Input
    id="id"
    onChange={[MockFunction]}
    question="question"
    value="hello world"
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="id"
      >
        question
      </label>
      <input
        autoComplete="off"
        className="form-control"
        id="id"
        onChange={[Function]}
        type="text"
        value="hello world"
      />
    </div>
  </Input>
</ControlSwitch>
`;
